ext/bcmath/libbcmath/src/div.c

```diff
@@ -31,219 +31,428 @@
+	 * 999 / 1000 = 0, Numbers that are not divisible are carried forward to the next division.
+	 * 9990000 / 1000 = 9990 (Here, the temporary quotient is 9990 and the error E is 10.)
+	 *
+	 * If calculate the temporary quotient using only one array of n1 and n2, the error E can be larger than 1.
+	 * In other words, in the restoring division, the count of additions for restore increases significantly.
+	 *
+	 * Therefore, in order to keep the error within 1 and to limit the number of additions required for restoration to one
+	 * at most, adjust the number of high-order digits used to calculate the temporary quotient as follows.
+	 * - Adjust the number of digits of n2 used in the calculation to BC_VECTOR_SIZE + 1 digit. The missing digits are
+	 *   filled in from the next array element.
+	 * - Add digits to n1 in the same way as the number of digits adjusted by n2.
+	 *
+	 * e.g. n1 = 123456780000, n2 = 123456789, radix = 10000
+	 * n1_arr = [0, 5678, 1234], n2_arr = [6789, 2345, 1]
+	 * n1_top = 1234, n2_top = 1
+	 * n2_top_tmp = 12345 (+4 digits), n1_top_tmp = 12345678 (+4 digits)
+	 * tmp_quot = n1_top_tmp / n2_top_tmp = 1000, tmp_rem = -9000
+	 * tmp_quot is too large, so tmp_quot--, tmp_rem += n2 (restoring)
+	 * quot = 999, rem = 123447789
+	 *
+	 * Details:
+	 * Suppose that when calculating the temporary quotient, only the high-order elements of the BC_VECTOR array are used.
+	 * At this time, n1 and n2 can be considered to be decomposed as follows. (Here, B = 10^b, any b ∈ ℕ)
+	 * n1 = n1_high * B^k + n1_low
+	 * n2 = n2_high * B^k + n2_low
+	 *
+	 * The error E can be expressed by the following formula.
+	 *
+	 * E = (n1_high * B^k) / (n2_high * B^k) - (n1_high * B^k + n1_low) / (n2_high * B^k + n2_low)
+	 * E = n1_high / n2_high - (n1_high * B^k + n1_low) / (n2_high * B^k + n2_low)
+	 * E = (n1_high * (n2_high * B^k + n2_low) - (n1_high * B^k + n1_low) * n2_high) / (n2_high * (n2_high * B^k + n2_low))
+	 * E = (n1_high * n2_low - n2_high * n1_low) / (n2_high * (n2_high * B^k + n2_low))
+	 *
+	 * Find the error MAX_E when the error E is maximum.
+	 * First, n1_high, which only exists in the numerator, uses its maximum value. Considering carry-back,
+	 * n1_high can be expressed as follows.
+	 * n1_high = n2_high * B
+	 * And n1_low is also present only in the numerator, but since it is a negative number, use the largest possible value, i.e. 0.
+	 * n1_low = 0
+	 *
+	 * MAX_E = (n1_high * n2_low - n2_high * n1_low) / (n2_high * (n2_high * B^k + n2_low))
+	 * MAX_E = (n2_high * B * n2_low) / (n2_high * (n2_high * B^k + n2_low))
+	 *
+	 * n2_low uses the maximum value.
+	 * n2_low = B^k - 1
+	 * MAX_E = (n2_high * B * n2_low) / (n2_high * (n2_high * B^k + n2_low))
+	 * MAX_E = (n2_high * B * (B^k - 1)) / (n2_high * (n2_high * B^k + B^k - 1))
+	 *
+	 * n2_high uses the minimum value, but want to see how the number of digits affects the error, so represent
+	 * the minimum value as:
```

Comment:
What minimum value is being used? We've been talking about maximum the whole time.

Reply:
Here, the smaller `n2_high` is, the larger `MAX_E` is. So use the minimum value.